internal/pkg/gateway/server/server.go

```
@@ -0,0 +1,53 @@
+/*
+Copyright 2020 IBM All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package server
+
+import (
+	pb "github.com/hyperledger/fabric-protos-go/gateway"
+	"github.com/hyperledger/fabric-protos-go/orderer"
+	"github.com/hyperledger/fabric-protos-go/peer"
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/hyperledger/fabric/discovery"
+	"github.com/hyperledger/fabric/internal/pkg/gateway/registry"
+	"google.golang.org/grpc"
+)
+
+var logger = flogging.MustGetLogger("gateway")
+
+// Server represents the GRPC server for the Gateway
+type Server struct {
+	registry Registry
+}
+
+// Registry represents the current network topology
+type Registry interface {
+	GetEndorsers(channel string, chaincode string) []peer.EndorserClient
+	GetDeliverers(channel string) []peer.DeliverClient
+	GetOrderers(channel string) []orderer.AtomicBroadcast_BroadcastClient
+	ListenForTxEvents(channel string, txid string, done chan<- bool) error
+}
+
+// NewGatewayServer creates a server side implementation of the gateway server grpc
+func newGatewayServer(registry Registry) (*Server, error) {
+	result := &Server{
+		registry,
+	}
+	return result, nil
+}
+
+func RegisterGatewayService(server *grpc.Server, discovery *discovery.Service) error {
+	registry, err := registry.NewRegistry(discovery)
+	if err != nil {
+		logger.Errorf("failed to create network registry: %s", err)
+		return err
+	}
+	gwServer, _ := newGatewayServer(registry)
```

ignoring errors is never a good idea.